FIX: plots.image show labels for every row if provided

julesvanrie opened the pull request:
## Overview

Closes #4112

Description of the changes proposed in this pull request:

`plots.image()` currently only displays labels for the first row, even if labels for every row (every image in the sample) have been provided. This breaks the use case in the [Explain ResNet50 using the Partition explainer](https://shap.readthedocs.io/en/latest/example_notebooks/image_examples/image_classification/Explain%20ResNet50%20using%20the%20Partition%20explainer.html) example. The use case is where the classes shown in the plot are not the same for every row (e.g. when one wants to only display the `top_k` classes, which would be different for every row).

This PR fixes this problem so that `plots.image()` displays:
- Labels for every row if they are provided.
- Only labels for the first row if only set of labels is provided.

This PR also adds a test case checking that `plots.image()` does not error when labels are provided in different formats that should be allowed to work.


## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

julesvanrie committed:
plots.image show labels for all rows if provided

Fixes https://github.com/shap/shap/issues/4112

First fix the labels reshape in case the labels are provided as a list.
Currently this creates an array of shape (1, nb_of_provided_labels).

If a list of list of labels is provided (# samples, # top_k classes),
this results in shape (1, # samples x # top_k classes), which does not
work when adding the labels to the plot later on.

This commit changes the reshaping logic to have labels with shape
(# samples, # top_k classes).
In case only one set of labels are provided in the list, the resulting
shape will be (1, # top_k classes).

Then we fix the actual plotting, where now only labels are added for the
first row, even if labels per row are provided (e.g. when using the
labels straight from the explainer, or when providing a 2D array, or a
list of lists). Thanks to the reshaping fix all these cases work.

Users can now decide to use the Explainer's labels, or provided them
manually with this flexibility:
- Separate labels for each sample (in the form of a 2D array, a list of
lists, or a flat list)
- One set of labels to be used for all samples (in the form of a flat
list, a 1D array, a 2D array (with first dim 1), a list containting 1
list).

Only requirement is that the number of labels provided is either:
- # samples x # top_k classes
- 1 x # top_k classes

Cases that have been tested, using the example mentioned in #4112 (in
turn a very slightly adapted version of the ResNet 50 tutorial):

Producing labels for every row:
```
shap.image_plot(shap_values)
shap.image_plot(shap_values, labels=[['a','b','c'],
                                     ['d','e','f']])
shap.image_plot(shap_values, labels=np.array([['a','b','c'],
                                              ['d','e','f']]))
shap.image_plot(shap_values, labels=['a','b','c','d','e','f'])
shap.image_plot(shap_values, labels=np.array(['a','b','c','d','e','f']))
```

Producing labels only for the first row:
(This was the intention of https://github.com/shap/shap/pull/3636)
```
shap.image_plot(shap_values, labels=['a', 'b', 'c'])
shap.image_plot(shap_values, labels=np.array(['a','b','c']))
shap.image_plot(shap_values, labels=[['a', 'b', 'c']])
shap.image_plot(shap_values, labels=np.array([['a','b','c']]))
```

julesvanrie committed:
test plots.image when labels are (not) provided

Test `image_plot()` for use cases where:
- User does not provide labels (use the ones from the Explanation)
- User provides labels for each image
- User only provides one set of labels for all images

julesvanrie committed:
type annotation for case of list of list of labels

Support the use case where labels for each image are provided as a list
of list of str.

julesvanrie committed:
changes and type annotations to pass mypy

CloseChoice commented on tests/plots/test_image.py:
```diff
@@ -89,3 +89,49 @@ def __init__(self, data, values, output_names):
     shap.plots.image_to_text(shap_values_test)
+
+
+def test_image_plot_with_labels():
```
Thanks for the test, this looks great. Would it be possible to set the seed for numpy and then use pytest-mpl to compare against the image?

Here is a blueprint: [test](https://github.com/shap/shap/blob/master/tests/plots/test_force.py#L94) and [baseline image](https://github.com/shap/shap/blob/master/tests/plots/baseline/test_force_plot_negative_sign.png).

julesvanrie replied:
Thanks for the review and the suggestion.

I have actually rewritten the tests to largely reuse the existing logic and example explanation of `test_image_multi` (by moving it into a fixture) and compare against baselines images that way.

CloseChoice reviewed (changes requested):
Thanks for the fix, that looks great. Would be great if you could update the test, then we're good to merge.

CloseChoice commented:
nevermind about the test failures, those are just connection issues when we try to download the dataset

julesvanrie committed:
Refactor creation of explanation into a fixture

To be able to reuse the created explanation object in other tests.

Also corrected the `output_names` in the explanation object which should
have as many elements as there are classes, not images.

julesvanrie committed:
tests compare against baseline image with labels

Change the tests so they can compare against baseline images instead of
merely checking they do not produce an error.

Reused the existing logic from `test_image_multi` thanks to refactoring
the creation of the explanation object into a fixture in the previous
commit.

This introduces 3 new image comparisons:
- No labels are provided (the ones from the explanation
are then used by default, first row only)
- Different labels per row are provided as a list of lists
- Different labels per row are provided as a np.array

The existing test case already covered:
- One set of labels is provided as a list

CloseChoice committed:
Merge branch 'master' into fix/plots-image-labels-multiple-rows

CloseChoice reviewed (approved):
looks great, ready to merge when pipelines succeed! Thanks a lot, awesome work
